doc/rsp.md

```diff
@@ -220,32 +220,42 @@ that the 32 registers are logically divided into 4 groups (0-7, 8-15, 16-23, 24-
 
 The lanes affected within the register group are laid out in *diagonal* layout; for
 instance, if `vt` is zero, the lanes will be: `VREG[0]<0>`, `VREG[1]<1>`, ...,
-`VREG[7]<7>`. `element(3..1)` specifies the lane affected in the first register of
-the register group, and thus identifies the diagonal (`element(0)` is ignored). 
+`VREG[7]<7>`. `element(3..1)` specifies the first register affected within the
+register group, and thus identifies the diagonal (`element(0)` is ignored).
 
 The following table shows the numbering of the 8 diagonals present in a 8-registers
 group; each cell of the table contains the diagonal that lane belongs to:
 
 | Reg | Lane 0 | Lane 1 | Lane 2 | Lane 3 | Lane 4 | Lane 5 | Lane 6 | Lane 7 |
-| --- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
-| `v0` | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
-| `v1` | 7 | 0 | 1 | 2 | 3 | 4 | 5 | 6 |
-| `v2` | 6 | 7 | 0 | 1 | 2 | 3 | 4 | 5 |
-| `v3` | 5 | 6 | 7 | 0 | 1 | 2 | 3 | 4 |
-| `v4` | 4 | 5 | 6 | 7 | 0 | 1 | 2 | 3 |
-| `v5` | 3 | 4 | 5 | 6 | 7 | 0 | 1 | 2 |
-| `v6` | 2 | 3 | 4 | 5 | 6 | 7 | 0 | 1 |
-| `v7` | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 0 |
-
-The address in DMEM from which the first lane is read/write is `GPR[base] + offset*16`;
-following lanes are read/written from subsequent memory addresses, wrapping around at the
-second 64-bit boundary. For instance, `LTV v0[e0],$1E(r0)` reads the lanes from the
-following addresses: `$1E`, `$20`, `$22`, `$24`, `$26`, `$18`, `$1C`.
-
-By combining `LTV` and `STV`, it is possible to transpose a matrix because diagonals
+| --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| `v0` | 0 | 7 | 6 | 5 | 4 | 3 | 2 | 1 |
+| `v1` | 1 | 0 | 7 | 6 | 5 | 4 | 3 | 2 |
+| `v2` | 2 | 1 | 0 | 7 | 6 | 5 | 4 | 3 |
+| `v3` | 3 | 2 | 1 | 0 | 7 | 6 | 5 | 4 |
+| `v4` | 4 | 3 | 2 | 1 | 0 | 7 | 6 | 5 |
+| `v5` | 5 | 4 | 3 | 2 | 1 | 0 | 7 | 6 |
+| `v6` | 6 | 5 | 4 | 3 | 2 | 1 | 0 | 7 |
+| `v7` | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
+
+`STV` writes lane 0 of the specified diagonal to the address `GPR[base] + (offset * 16)`;
+following lanes are written to subsequent memory addresses, wrapping around at the
+second 64-bit boundary. For instance, `STV v0[e2],$1E(r0)` writes diagonal 1, starting
+with `VPR[1]<0>`, to the following addresses: `$1E`, `$20`, `$22`, `$24`, `$26`, `$18`,
+`$1A`, `$1C`.
+
+`LTV` reads 128 bits from `(GPR[base] + (offset * 16)) & 0xFF8`, rotates them left by
+`element + (address & 0x8)` bytes, then loads the leftmost 16 bits into lane 0 of the
+specified diagonal, the next 16 bits into lane 1, and so on. If `element(0)` is 0
+and the address is 128-bit aligned, this has the effect of reading the given diagonal
+starting from register `v0` of the register group. For instance, `LTV v0[e2],$1E(r0)`
+reads diagonal 1, starting with `VPR[0]<7>`, from the following addresses: `$20`, `$22`,
+`$24`, `$26`, `$18`, `$1A`, `$1C`, `$1E`.
```

Comment:
I don't specifically like this description because it is closer to how one would write an emulator for it, than a description of what the hardware does. I know it is a weird opcode, but I think we should strive to describe the hardware.

This is how I understand it:

```suggestion
`LTV` fetches two subsequent 64-bit aligned words, starting from `(GPR[base] + (offset * 16)) & 0xFF8`, and combine them into a 128-bit intermediate value called `W`, that will be used to populate the diagonal. The highest half in `W` is populated with the only 64-bit word among the two whose address is also 128-bit aligned. Lane 0 of the diagonal is then populated with the 16 bits in `W `starting at byte offset `element(3..0)` (`W[element(3..0)..element(3..0)+1]`). Subsequent lanes are populated with subsequent 16 bits words in `W`, wrapping around within `W`. For instance `LTV v0[e3],$1E(r0)` reads diagonal 1, starting with `VPR[0]<7>`, from `DMEM[$23..$24]`, `DMEM[$25..$26]`, `DMEM[$27,$18]`, `DMEM[$19..$1A]`, `DMEM[$1B..$1C]`, `DMEM[$1D..$1E]`, `DMEM[$1F..$20]`, `DMEM[$21..$22]`
```

(also notice I think your example was incorrect, I think I've fixed it but please double-check).

Reply:
The suggestion is great and definitely less confusing. I think my example looks a little different because it starts with `VPR[0]<7>`, rather than `VPR[1]<0>`. In your suggestion did you meant to start with `VPR[1]<0>`? If so I think we are in agreement here.

Reply:
Uhm no, I would say that `VPR[0]<7> = DMEM[$23..$24]`. It looks like in your example (adjusted with `e3` instead of `e2`) you wrote that `VPR[0]<7> = DMEM[$21..$22]`. 

This is my understanding. Offset is $1E. The two 64-bit words are $18-1F and $20-$27. The second one is the 128-bit aligned one, so `W = DMEM[$20..$27,$18..$1F]`. `e3` correspond to an offset of 3 bytes, so the first 16-bit word is `W[3..4] = DMEM[$23..$24]`. `e3` also means diagonal 1; lane 0 of diagonal 1 is `<7>`, so the the first value `DMEM[$23..$24]` is written to `VPR[0]<7>`.

Reply:
I used the notation `VPR[reg]<lane>`, so lane 0 of any diagonal is `VPR[diagonal]<0>`. As I understand it the value loaded into lane 0 stays the same for even `element` values, so that when you transpose a matrix the starting address you give is loaded into the appropriate lane of register 0, mirroring how the starting address you give `STV` is stored from lane 0 of the appropriate register.